modal/_container_entrypoint.py

```diff
@@ -962,22 +944,49 @@ def breakpoint_wrapper():
         if imp_fun.obj is not None and not imp_fun.is_auto_snapshot:
             post_checkpoint_methods = _find_callables_for_obj(imp_fun.obj, _PartialFunctionFlags.ENTER_POST_CHECKPOINT)
-            event_loop.run(call_functions_for_setup(function_io_manager, post_checkpoint_methods.values()))
+            event_loop.run(call_lifecycle_functions(function_io_manager, post_checkpoint_methods.values()))
 
-        if not imp_fun.is_async:
-            call_function_sync(function_io_manager, imp_fun)
-        else:
-            event_loop.run(call_function_async(function_io_manager, imp_fun))
+        # Execute the function.
+        try:
+            if imp_fun.is_async:
+                event_loop.run(call_function_async(function_io_manager, imp_fun))
+            else:
+                # Set up a signal handler for `SIGUSR1`, which gets translated to an InputCancellation
+                # during function execution. This is sent to cancel inputs from the user.
+                def _cancel_input_signal_handler(signum, stackframe):
+                    raise InputCancellation("Input was cancelled by user")
 
-        # Commit on exit to catch uncommitted volume changes and surface background
-        # commit errors.
-        function_io_manager.volume_commit(
-            [v.volume_id for v in container_args.function_def.volume_mounts if v.allow_background_commits]
-        )
-        function_io_manager.stop_heartbeat()  # avoid "Canceling remaining unfinished task" warnings etc.
+                signal.signal(signal.SIGUSR1, _cancel_input_signal_handler)
+
+                call_function_sync(function_io_manager, imp_fun)
+        finally:
+            # Run exit handlers. From this point onward, ignore all SIGINT signals that come from
+            # graceful shutdowns originating on the worker, as well as stray SIGUSR1 signals that
+            # may have been sent to cancel inputs.
+            int_handler = signal.signal(signal.SIGINT, signal.SIG_IGN)
+            usr1_handler = signal.signal(signal.SIGUSR1, signal.SIG_IGN)
+
+            try:
+                # Identify "exit" methods and run them.
+                if imp_fun.obj is not None and not imp_fun.is_auto_snapshot:
+                    exit_methods = _find_callables_for_obj(imp_fun.obj, _PartialFunctionFlags.EXIT)
+                    event_loop.run(call_lifecycle_functions(function_io_manager, exit_methods.values()))
+
+                # Finally, commit on exit to catch uncommitted volume changes and surface background
+                # commit errors.
+                function_io_manager.volume_commit(
+                    [v.volume_id for v in container_args.function_def.volume_mounts if v.allow_background_commits]
+                )
+                # Avoid "Canceling remaining unfinished task" warnings.
+                function_io_manager.stop_heartbeat()
+
+            finally:
+                # Restore the original signal handler, needed for container_test hygiene.
```

Comment:
This has to do with us running multiple tests of the container in the same process I assume?

Reply:
Yes, I updated the comment to be clearer here